doc/kernel-limitations.md

```diff
@@ -0,0 +1,174 @@
+process specific information, including `pid`, `ppid`, `fd table`, etc.
+
+This approach would allow child processes created with `posix_spawn` to
+run alongside the parent process. Sure, there is no memory address space
+separation between the processes, but the child or the parent would have
+no handle to reference the other side's memory.
+
+On the other hand, a child process created with `fork` is supposed to
+share variables with the parent. It's significantly more challenging to
+emulating a forked child process as a thread without underlying support
+of address space separation.
+
+
+## Limitations arisen from SGX's thread model
+
+With SGX1, the number of threads running inside an enclave, i.e.,
+**ethread**, has to be statically declared. Once the application starts,
+no more ethreads can be dynamically created. Meanwhile, the stack size
+of each ethread is also statically declared and allocated while creating
+the enclave. An application attempting to create more ethreads than what
+is declared will crash.
+
+A Mystikos application declares how many user threads are needed in
+config.json. A potential mitigation of the limitation is to make the setting big
+enough. The caveat is that we should ensure the stack size for each ethread
+is reasonable so we don't waste too much stack memory on unused threads.
+
+When an user application calls `pthread_create`, Mystikos makes an OCALL
+to the host to create a host thread, and re-enters the enclave with an
+ECALL, and once inside the enclave, jump to the thread routine. The long
+chain of events leads to delays and performance problems occasionally.
+
+The security implication of using host threads is that the untrusted host
+controls the thread scheduling, and may decide to starve or switch execution
+order of ethreads.
+
+The security and performance impact of this limitation can be mitigated with
+an enclave-resident thread scheduler, which uses M:N mapping to map ethreads
+to user application pthreads, and performs context switching between pthreads.
+
+This limitation is likely to be relaxed when EDMM of SGX2 is officially
+supported in Mystikos.
+
+## Limitations arisen from SGX's memory model
+
+With SGX1, the heap size of the enclave application has to be statically
+declared. Once the application starts, allocating more memory than what's
+declared causes a crash due to out of memory.
+
+A Mystikos application declares how big the heap size is needed in config.json.
+A potential mitigation of the limitation is to make the setting big enough.
+The downside of doing that is: 1) the SGX runtime will take more time to
+add/extend heap pages one by one during the initialization, and 2) The heap
+size is backed by physical EPC memory of the system. If the declared heap
+size is much more than the available physical EPC memory size, the application
+performance is likely to suffer due to severe `EPC paging`.
+
+Also with SGX1, the untrusted host can manipulate the page permissions at
```

Comment:
```suggestion
Also with SGX1, the Enclave cannot change Enclave enforced page access restriction after the Enclave is initialized. It's possible for the enclave SW to request the host to enforce the restriction in the page tables, but with the SGX thread model, the Enclave SW should assume the host might silently drop the requests.
```

Reply:
Untrusted privileged system SW involved in MMU control, for example, OS page table manger, VMM EPT manager can always manipulate the page permission in the construct it controls and trigger page fault. Protection against DoS attack from the privileged system SW is typically outside of the scope of TEE Technologies. This is true for Intel SGX, as well as AMD SEV. The security objective is that the TEE can implement its own page permission restriction, even if the untrusted privileged system SW does not enforce the policy (TEE's page permission policy is a subset of the untrusted policy). Even if the untrusted VMM (for both Intel SGX and AMD SEV) does not enforce any page access permission policy, with Intel SGX, the Enclave can enforce its own page permission policy in SGX EPCM(OS page table policy is untrusted for SGX too) , and with AMD SEV, the trusted kernel in a SEV protected VM can enforce its own page permission policy through OS page table. The result is that the trusted page access restriction maintained by the TEE is honored by the MMU, not the potentially more relaxed policy set by the untrusted system SW. The feature limitation of SGX1 is that the Enclave cannot change the Enclave enforced page access restriction after the Enclave is initialized. SGX2.0 does not have this limitation.